File: ArashiDNS.WhoisWorker/src/detection/availability-detector.ts
export type AvailabilityStatus = 'unknown' | 'not_registered' | 'registered' | 'reserved' | 'throttled' | 'error';

export interface AvailabilityResult {
  status: AvailabilityStatus;
  isAvailable: boolean;
  isThrottled: boolean;
  isReserved: boolean;
  errorMessage?: string;
}

const NOT_REGISTERED_PATTERNS = [
  /No match for\s+"/i,
  /No match for\s+domain/i,
  /^NOT FOUND/im,
  /^No Data Found/im,
  /Domain not found/i,
  /Status:\s*free\b/i,
  /Domain Status:\s*available\b/i,
  /^Status:\s*available$/im,
  /No such domain/i,
  /The queried object does not exist/i,
  /Domain not registered/i,
  /is free/i,
  /NOT EXIST/i,
  /No match for/i,
  /No records matching/i,
  /DOMAIN NOT FOUND/i,
  /^Not found$/im,
  /AVAILABLE FOR REGISTRATION/i,
  /No entries found/i,
  /^AVAILABLE$/im,
  /is available for registration/i,
  /queried domain name is not registered/i,
  /has not been registered/i,
  /no matching record/i,
  /is free for registration/i,
];

const REGISTERED_PATTERNS = [
  /Domain Name:/i,
  /Domain name:/i,
  /domain:/i,
  /Registry Domain ID:/i,
  /Registrar:/i,
  /Creation Date:/i,
  /Created:/i,
  /Registration Date:/i,
  /Registry Expiry Date:/i,
  /Expiration Date:/i,
  /Updated Date:/i,
  /Last Updated:/i,
  /Name Server:/i,
  /Nameserver:/i,
  /nserver:/i,
  /Domain Status:/i,
  /Status:/i,
];

const THROTTLED_PATTERNS = [
  /exceeded/i,
  /rate limit/i,
  /try again later/i,
  /too many requests/i,
  /query limit/i,
  /throttl/i,
  /abuse/i,
  /Please try again/i,
];

const RESERVED_PATTERNS = [
  /reserved/i,
  /reserved by/i,
  /reserved domain/i,
  /premium domain/i,
  /restricted/i,
];

export function detectAvailability(rawResponse: string | null): AvailabilityResult {
  if (!rawResponse) {
    return { status: 'unknown', isAvailable: false, isThrottled: false, isReserved: false, errorMessage: 'Empty response' };
  }

  if (THROTTLED_PATTERNS.some(p => p.test(rawResponse))) {
    return { status: 'throttled', isAvailable: false, isThrottled: true, isReserved: false };
  }

  if (RESERVED_PATTERNS.some(p => p.test(rawResponse))) {
    return { status: 'reserved', isAvailable: false, isThrottled: false, isReserved: true };
  }

  const isNotRegistered = NOT_REGISTERED_PATTERNS.some(p => p.test(rawResponse));
  if (isNotRegistered) {
    const hasRegisteredSignals = REGISTERED_PATTERNS.some(p => p.test(rawResponse));
    if (!hasRegisteredSignals) {
      return { status: 'not_registered', isAvailable: true, isThrottled: false, isReserved: false };
    }
  }

  const registeredCount = REGISTERED_PATTERNS.filter(p => p.test(rawResponse)).length;
  if (registeredCount >= 2) {
    return { status: 'registered', isAvailable: false, isThrottled: false, isReserved: false };
  }

  return { status: 'unknown', isAvailable: false, isThrottled: false, isReserved: false };
}

export function detectAvailabilityFromRdap(rawJson: string | null): AvailabilityResult {
  if (!rawJson) {
    return { status: 'unknown', isAvailable: false, isThrottled: false, isReserved: false, errorMessage: 'Empty RDAP response' };
  }

  try {
    const root = JSON.parse(rawJson);

    if (root.errorCode) {
      const code = root.errorCode;
      if (code === 404 || code === 410) {
        return { status: 'not_registered', isAvailable: true, isThrottled: false, isReserved: false };
      }
      if (code === 429) {
        return { status: 'throttled', isAvailable: false, isThrottled: true, isReserved: false };
      }
    }

    if (root.ldhName || root.unicodeName) {
      return { status: 'registered', isAvailable: false, isThrottled: false, isReserved: false };
    }
  } catch {
    // Not valid JSON
  }

  return { status: 'unknown', isAvailable: false, isThrottled: false, isReserved: false };
}

File: ArashiDNS.WhoisLib/Parsing/AvailabilityDetector.cs
using System.Text.Json;
using System.Text.RegularExpressions;
using ArashiDNS.WhoisLib.Contracts.Models;

namespace ArashiDNS.WhoisLib.Parsing;

/// <summary>
/// Detects domain availability status from WHOIS/RDAP responses
/// </summary>
public partial class AvailabilityDetector
{
    private static readonly string[] NotRegisteredPatterns =
    [
        @"No match for\s+""",
        @"No match for\s+domain",
        @"^NOT FOUND",
        @"^No Data Found",
        @"Domain not found",
        @"Status:\s*free\b",
        @"Domain Status:\s*available\b",
        @"^Status:\s*available$",
        @"No such domain",
        @"The queried object does not exist",
        @"Domain not registered",
        @"is free",
        @"NOT EXIST",
        @"No match for",
        @"No records matching",
        @"DOMAIN NOT FOUND",
        @"^Not found$",
        @"AVAILABLE FOR REGISTRATION",
        @"No entries found",
        @"^AVAILABLE$",
        @"is available for registration",
        @"queried domain name is not registered",
        @"has not been registered",
        @"no matching record",
        @"is free for registration"
    ];

    private static readonly string[] RegisteredPatterns =
    [
        @"Domain Name:",
        @"Domain name:",
        @"domain:",
        @"Registry Domain ID:",
        @"Registrar:",
        @"Creation Date:",
        @"Created:",
        @"Registration Date:",
        @"Registry Expiry Date:",
        @"Expiration Date:",
        @"Updated Date:",
        @"Last Updated:",
        @"Name Server:",
        @"Nameserver:",
        @"nserver:",
        @"Domain Status:",
        @"Status:"
    ];

    private static readonly string[] ThrottledPatterns =
    [
        @"exceeded",
        @"rate limit",
        @"try again later",
        @"too many requests",
        @"query limit",
        @"throttl",
        @"abuse",
        @"Please try again"
    ];

    private static readonly string[] ReservedPatterns =
    [
        @"reserved",
        @"reserved by",
        @"reserved domain",
        @"premium domain",
        @"restricted"
    ];

    private static readonly string[] ServerSpecificNotFounds = new[]
    {
        @"No match for ""{query}""\.",  // Verisign
        @"No match for domain",         // Generic
        @"{query} no match",            // Some ccTLDs
    };

    public AvailabilityResult Detect(string? rawResponse, WhoisQueryType queryType = WhoisQueryType.Domain)
    {
        if (string.IsNullOrWhiteSpace(rawResponse))
        {
            return new AvailabilityResult
            {
                Status = AvailabilityStatus.Unknown,
                IsAvailable = false,
                ErrorMessage = "Empty response"
            };
        }

        var lower = rawResponse.ToLowerInvariant();

        if (IsThrottled(rawResponse, lower))
        {
            return new AvailabilityResult
            {
                Status = AvailabilityStatus.Throttled,
                IsAvailable = false,
                IsThrottled = true
            };
        }

        if (IsReserved(rawResponse, lower))
        {
            return new AvailabilityResult
            {
                Status = AvailabilityStatus.Reserved,
                IsAvailable = false,
                IsReserved = true
            };
        }

        if (IsNotRegistered(rawResponse, lower))
        {
            return new AvailabilityResult
            {
                Status = AvailabilityStatus.NotRegistered,
                IsAvailable = true
            };
        }

        if (IsRegistered(rawResponse, lower))
        {
            return new AvailabilityResult
            {
                Status = AvailabilityStatus.Registered,
                IsAvailable = false
            };
        }

        return new AvailabilityResult
        {
            Status = AvailabilityStatus.Unknown,
            IsAvailable = false
        };
    }

    public AvailabilityResult DetectFromRdap(string? rawJson)
    {
        if (string.IsNullOrWhiteSpace(rawJson))
        {
            return new AvailabilityResult
            {
                Status = AvailabilityStatus.Unknown,
                IsAvailable = false,
                ErrorMessage = "Empty RDAP response"
            };
        }

        try
        {
            using var doc = JsonDocument.Parse(rawJson);
            var root = doc.RootElement;

            if (root.TryGetProperty("errorCode", out var errorCode))
            {
                var code = errorCode.GetInt32();
                if (code is 404 or 410)
                {
                    return new AvailabilityResult
                    {
                        Status = AvailabilityStatus.NotRegistered,
                        IsAvailable = true
                    };
                }
                if (code is 429)
                {
                    return new AvailabilityResult
                    {
                        Status = AvailabilityStatus.Throttled,
                        IsAvailable = false,
                        IsThrottled = true
                    };
                }
            }

            if (root.TryGetProperty("ldhName", out _) || root.TryGetProperty("unicodeName", out _))
            {
                return new AvailabilityResult
                {
                    Status = AvailabilityStatus.Registered,
                    IsAvailable = false
                };
            }

            if (root.TryGetProperty("notices", out var notices))
            {
                foreach (var notice in notices.EnumerateArray())
                {
                    var title = notice.TryGetProperty("title", out var t) ? t.GetString() : null;
                    if (title != null && title.Contains("Registration", StringComparison.OrdinalIgnoreCase)
                                     && title.Contains("Available", StringComparison.OrdinalIgnoreCase))
                    {
                        return new AvailabilityResult
                        {
                            Status = AvailabilityStatus.NotRegistered,
                            IsAvailable = true
                        };
                    }
                }
            }
        }
        catch
        {
            // JSON parse error - not a valid RDAP response
        }

        return new AvailabilityResult
        {
            Status = AvailabilityStatus.Unknown,
            IsAvailable = false
        };
    }

    private static bool IsNotRegistered(string raw, string lower)
    {
        foreach (var pattern in NotRegisteredPatterns)
        {
            if (Regex.IsMatch(raw, pattern, RegexOptions.IgnoreCase | RegexOptions.Multiline))
            {
                var hasRegisteredSignals = false;
                foreach (var regPattern in RegisteredPatterns)
                {
                    if (Regex.IsMatch(raw, regPattern, RegexOptions.IgnoreCase))
                    {
                        hasRegisteredSignals = true;
                        break;
                    }
                }

                if (!hasRegisteredSignals)
                    return true;
            }
        }

        return false;
    }

    private static bool IsRegistered(string raw, string lower)
    {
        var matchCount = 0;
        foreach (var pattern in RegisteredPatterns)
        {
            if (Regex.IsMatch(raw, pattern, RegexOptions.IgnoreCase))
            {
                matchCount++;
                if (matchCount >= 2) return true;
            }
        }
        return false;
    }

    private static bool IsThrottled(string raw, string lower)
    {
        foreach (var pattern in ThrottledPatterns)
        {
            if (Regex.IsMatch(raw, pattern, RegexOptions.IgnoreCase))
                return true;
        }
        return false;
    }

    private static bool IsReserved(string raw, string lower)
    {
        foreach (var pattern in ReservedPatterns)
        {
            if (Regex.IsMatch(raw, pattern, RegexOptions.IgnoreCase))
                return true;
        }
        return false;
    }
}

public class AvailabilityResult
{
    public AvailabilityStatus Status { get; set; }
    public bool IsAvailable { get; set; }
    public bool IsThrottled { get; set; }
    public bool IsReserved { get; set; }
    public string? ErrorMessage { get; set; }
}

public enum AvailabilityStatus
{
    Unknown,
    NotRegistered,
    Registered,
    Reserved,
    Throttled,
    Error
}
